### ISSUE TITLE
[Feature]: Support disabling tsconfig paths

### ISSUE THREAD

**Issue by chenjiahan** [web-infra-dev/modern.js#4192]
### What problem does this feature solve?

Currently, the Modern.js Builder automatically reads the `paths` field in the `tsconfig.json` as the alias config, and the priority of tsconfig paths is higher than `source.alias`, so the `source.alias` is useless for TypeScript projects.

But in some scenarios, the user does not want the default behavior, they want to use `source.alias` because it allows user to create some alias config dynamically, while tsconfig paths is static and not as flexible as `source.alias`. In this case, tsconfig paths is only used to provide the type of alias, and will not override the user configured alias config.

### What does the proposed API look like?

The Modern.js Builder will add a new `source.ignoreTsConfigPaths` config to disable the default behavior, so users can use the `source.alias` and it will not be overridden by tsconfig paths.

```ts
export default {
  source: {
    ignoreTsConfigPaths: true,
  }
}
```

**Comment by fireairforce** [comment]
The specific implementation is to disable `TsConfigPathsPlugin` after the user configures `source.ignoreTsConfigPaths` as true?

**Comment by h-a-n-a** [comment]
Great RFC! Personally speaking, consistency with the original configuration is import. Since `source.alias` is known for the most of modern.js users and `tsConfigPath` is an implicit behavior, adding the new field may make user a little bit confused. I would love to use the `alias` related field to at least show the connection between the original field.

Secondly, I don't know if we will have other fields or configuration files that will be supported natively by modern.js. 

Based on my considerations, I would like to propose something like `source.aliasStrategy`. The default value can be `'prefer-tscconfig'`. For the extendability, users can pass array to specify the order of configurations in the future.

**Comment by chenjiahan** [comment]
> The specific implementation is to disable TsConfigPathsPlugin after the user configures source.ignoreTsConfigPaths as true?

Yes, and when using Rspack, it will remove the [resolve.tsConfigPaths](https://www.rspack.dev/config/resolve.html#resolvetsconfigpath) config.

**Comment by chenjiahan** [comment]
> I would like to propose something like source.aliasStrategy. The default value can be 'prefer-tsconfig'

Like this idea! I will update my RFC and use the `source.aliasStrategy`.

> for the most of modern.js users and tsConfigPath is an implicit behavior

Considering that next.js reads tsconfig paths by default, maybe community users are more familiar with tsconfig paths rather than source alias.

> I don't know if we will have other fields or configuration files that will be supported natively by modern.js.

Only tsconfig paths now. 😄

**Comment by KyrieLii** [comment]
Gread idea ! In addition, maybe we need to provide a configuration to specify the tsconfig.json file path.  

The current implementation is hard-coded `${rootPath}/tsconfig.json`.

**Comment by KyrieLii** [comment]
> find the closet tsconfig.json of each module

What I mean is that the following configuration can be directly modified:  
(Although we can now modify it through `tools.bundlerChain`.)

```ts
// webpack
new TsconfigPathsPlugin({
  configFile: 'xxx' // here
})
```

```ts
// rspack
resolve: {
    tsConfigPath: 'xxx', // here
  },
```

For example, some projects have diffrent `tsconfig.web.json`, `tsconfig.node.json` or `tsconfig.dev.json`, `tsconfig.prod.json`. 😅

**Comment by hardfist** [comment]
> > The specific implementation is to disable TsConfigPathsPlugin after the user configures source.ignoreTsConfigPaths as true?
> 
> Yes, and when using Rspack, it will remove the [resolve.tsConfigPaths](https://www.rspack.dev/config/resolve.html#resolvetsconfigpath) config.

things becomes complex now, since rspack tsconfig paths and alias are merged and tsconfig path has high priority, so it would be hard for rspack to implement the stragegy

**Comment by chenjiahan** [comment]
> What I mean is that the following configuration can be directly modified

If the user asks for this config, we will consider adding it.

**Comment by chenjiahan** [comment]
> things becomes complex now, since rspack tsconfig paths and alias are merged and tsconfig path has high priority, so it would be hard for rspack to implement the stragegy

We can just remove the [resolve.tsConfigPaths](https://www.rspack.dev/config/resolve.html#resolvetsconfigpath) config fro Modern.js Builder when using `aliasStrategy:'prefer-tsconfig'`.

**Comment by hardfist** [comment]
> We can just remove the [resolve.tsConfigPaths](https://www.rspack.dev/config/resolve.html#resolvetsconfigpath) config fro Modern.js Builder when using aliasStrategy:'prefer-tsconfig'.

you mean `perfer-alias`?

**Comment by chenjiahan** [comment]
> you mean perfer-alias?

Yep, my mistake

**Comment by hardfist** [comment]
it seems rspack merge tsconfig.paths and alias if set tsconfigpaths which behaves not the same as prefer-tsconfig or prefer-alias

**Comment by chenjiahan** [comment]
> rspack merge tsconfig.paths and alias

If the tsconfig.paths has a higher priority, then is it same as `prefer-tsconfig`.

**Comment by chenjiahan** [comment]
The Rspack documentation should explain more about the relationship between `tsconfig.paths` and `resolve.alias`.

**Comment by hardfist** [comment]
@bvanjoi @Boshen can you help check the result and help improve the documentation of resolve
